core/dbt/parser/schemas.py

```diff
@@ -1005,6 +1056,76 @@ def _target_type(self) -> Type[UnparsedNodeUpdate]:
+            assert isinstance(self.yaml.file, SchemaSourceFile)
+            source_file: SchemaSourceFile = self.yaml.file
+            latest_version = target.latest_version or max(versions).v
+            for unparsed_version in versions:
+                versioned_model_name = (
+                    unparsed_version.defined_in or f"{block.name}_{unparsed_version.formatted_v}"
+                )
+                # ref lookup without version - version is not set yet
+                versioned_model_unique_id = self.manifest.ref_lookup.get_unique_id(
+                    versioned_model_name, None, None
+                )
+                if versioned_model_unique_id is None:
+                    # TODO: handle disabled nodes
+                    warn_or_error(
+                        NoNodeForYamlKey(
+                            patch_name=versioned_model_name,
+                            yaml_key=target.yaml_key,
+                            file_path=source_file.path.original_file_path,
+                        )
+                    )
+                    continue
+                versioned_model_node = self.manifest.nodes.pop(versioned_model_unique_id)
+                versioned_model_node.unique_id = (
+                    f"model.{target.package_name}.{target.name}.{unparsed_version.formatted_v}"
+                )
+                versioned_model_node.fqn[-1] = target.name
+                versioned_model_node.fqn.append(unparsed_version.formatted_v)
+                self.manifest.add_node_nofile(versioned_model_node)
+
+                # flatten columns based on include/exclude
+                version_refs: ParserRef = ParserRef.from_versioned_target(
+                    block.target, unparsed_version.v
+                )
+
+                versioned_model_patch = ParsedNodePatch(
+                    name=target.name,
+                    original_file_path=target.original_file_path,
+                    yaml_key=target.yaml_key,
+                    package_name=target.package_name,
+                    description=unparsed_version.description or target.description,
+                    columns=version_refs.column_info,
+                    meta={**target.meta, **unparsed_version.meta},
+                    docs=unparsed_version.docs or target.docs,
+                    config={**target.config, **unparsed_version.config},
+                    access=unparsed_version.access or target.access,
+                    version=unparsed_version.v,
+                    is_latest_version=latest_version == unparsed_version.v,
+                )
+                versioned_model_node.patch(versioned_model_patch)
+                self.patch_node_config(versioned_model_node, versioned_model_patch)
+                source_file.append_patch(
+                    versioned_model_patch.yaml_key, versioned_model_node.unique_id
+                )
+            self.manifest.rebuild_ref_lookup()
```

Comment:
TODO: look into whether this can be more granular and moved into the loop for a performance improvement